### DISCUSSION TITLE
Why does the Netlify template require to have at least `overrides/hooks/on_env.py` as hook and `generate_graph: false` as extra?

### DISCUSSION THREAD

**Discussion by ooker777** [ObsidianPublisher/obsidian-github-publisher#255] [Q&A — Template] [answered]
I notice the minimal workable `mkdocs.yaml` setting for the latest commit of the [netlify template](https://github.com/ObsidianPublisher/template-netlify-vercel/tree/main/docs) is:
```yaml
site_name: your_site_name

theme:
    name: 'material'
    custom_dir: overrides
hooks:
  - overrides/hooks/on_env.py
extra:
  generate_graph: false
```

# Error if the hook is not provided
```
  File "D:\Programming\Static web\mkdocs\template-netlify-vercel\overrides\main.html", line 27, in template   
    {% set pg_title = 'title' | value_in_frontmatter(page.meta) %}
    ^^^^^^^^^^^^^^^^^^^^^^^^^
jinja2.exceptions.TemplateAssertionError: No filter named 'value_in_frontmatter'.
```

# Error if the extra is not provided
```
  File "D:\Programming\Static web\mkdocs\template-netlify-vercel\overrides\hooks\on_env.py", line 181, in on_env
    obsidian_graph()
  File "D:\Programming\Static web\mkdocs\template-netlify-vercel\overrides\hooks\on_env.py", line 16, in obsidian_graph
    import obsidiantools.api as otools
```

The code works if `custom_dir` is not provided. Do you have any idea?

**Comment by Mara-Li** [comment, marked as answer]
Extra and hooks have different purpose.
Hooks run python script and is needed for the build.
Extra add custom yaml parameters needed to tell the on_env script how to build the website.

**Reply by ooker777** [reply]
why does the `generate_graph: false` config need to be explicit? Why not implicitly set it to false if absence?

**Reply by Mara-Li** [reply]
1. I will deprecate soon the netlify template (probably in January 2004), as Netlify refuse to upgrade their python version. Python 3.11 will be the minimum.
2. The GitHub template doesn't need this key. So the generating is True by default, as I can use python 3.11+ version.

**Reply by Mara-Li** [reply]
Moreover, in this Holiday season / after my finals, I will upgrade the template to have only one repository and set-up a scripts where you can define the value of `mkdocs.yaml` directly.
The idea is to run the Github Actions, fill the value, and you don't need extra configuration!

**Reply by ooker777** [reply]
As stated in [Compatibility of publishing solution](https://obsidian-publisher.netlify.app/Template/Compatibility/ "Compatibility of publishing solution"), GitHub Pages building time is very long. Is there a way to build it on my machine first and then push the result after automatically?

**Reply by Mara-Li** [reply]
Well, i don't think it will be better. 
Here is my timer :
![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/30244939/522b7ace-e0f1-4b14-aeb8-65ea133dd23f)

**Reply by Mara-Li** [reply]
I accelerated the build because : 
- Python 3.11+
- Optimization with assets

**Reply by Mara-Li** [reply]
A future upgrade of Material should accelerate more the build, but the plugin is paid for the moment.
